compiler/qsc_ast/src/ast.rs

```diff
@@ -109,12 +115,28 @@ impl Hash for NodeId {
 pub struct Package {
     /// The node ID.
     pub id: NodeId,
-    /// The namespaces in the package.
-    pub namespaces: Box<[Namespace]>,
+    /// The top-level syntax nodes in the package.
+    pub nodes: Box<[TopLevelNode]>,
     /// The entry expression for an executable package.
     pub entry: Option<Box<Expr>>,
 }
 
+fn take_and_concat<T>(left: &mut Box<[T]>, right: &mut Box<[T]>) -> Box<[T]> {
+    let mut v = std::mem::take(left).into_vec();
+    v.reserve_exact(right.len());
+    v.extend(std::mem::take(right).into_vec());
+    v.into_boxed_slice()
+}
+
+impl Package {
+    /// Extends the `Package` with the contents of another `Package`.
+    /// `other` should not contain any `NodeId`s
+    /// that conflict with the current `Package`.
```

Comment:
You can turn this into a check using the `Validator` from qsc_ast/src/validate.rs. It will walk the tree and panic if it finds any duplicate ids. I think that would be worth doing in this function after the call to `take_and_concat`.

Reply:
Ooh that would work perfectly, thanks!

Reply:
Huh. So I added the AST validator, which immediately caught the fact that the `NodeId` on the AST `Package` is never set! This got me thinking a little deeper about the semantics of `ast::Package::extend()`. Should it retain the current package ID? Should it adopt the Package ID of `other`? Come to think of it, shouldn't we be creating a new unique ID for the resulting package, since the result of `extend()` is a new, unique, AST? I really liked that idea. The same `NodeId` should never refer to two different "states" of a node. Those two "states" are actually two different trees, and trees should be immutable. Intuitively, I feel like this principle sets us up for better debuggability, and perhaps even things like partial recompilation, down the line.

So instead of keeping this `extend()` method that takes a mutable reference to self, I think I'll create a function with a name like `concat()`, which takes two `Package`s and creates a new, merged, one, with a new unique ID.

Reply:
Pushed commit to fix this: https://github.com/microsoft/qsharp/pull/762/commits/ea50f1b6f7bcab763269b83ec2573f4e1518e9da

I ended up pulling the implementation of `extend()`/`concat()` out of ast.rs and hir.rs. The layering seemed problematic - these functions are very specific to incremental compilation, and they now take a dependency on the validators.... so it seemed wrong that they would live alongside the AST/HIR struct definitions. They live with the incremental compiler now.